More clarity around validating proposals (#675)
mlswg/mls-protocol: draft-ietf-mls-protocol.md

  to add the client to the group
* When a LeafNode is received by a group member in an Add, Update, or Commit
  message
* When a client joining a group receives LeafNode objects for the leaves of
  of the group's ratchet tree

To verify that a LeafNode is valid regardless of its use, the client starts

> with the following steps:

Reviewer: I'm not clear why this says "regardless of its use".  Maybe you mean "In all these cases...", but I would prefer leaving it unadorned to not imply that there's a restriction to those cases (in case another one comes up).

If you're going to say "starts with..." you need to have more steps that follow.

Author: Thanks for pointing this out! (We used to have a earlier version with more cases and more steps.)

  another LeafNode, the authentication service MUST additionally validate that
  the set of identities attested by the credential in the new LeafNode is
  acceptable relative to the identities attested by the old credential.
  For example:
    * An Update proposal updates the sender's old LeafNode to a new one

>     * A "resync" external commit removes the joiner's old LeafNode via a Remove proposal and replaces it with a new one

Reviewer 2: Also checking if identities are the same in a resync is a good point, but isn't it undermined by doing the external join and remove separately?

Reviewer 3: I think this is covered in the External Commits section, but mentioning it here or including a reference probably doesn't hurt.

Author: Currently this requirement is in the External Commit section:
>  There MAY be a single Remove proposal, where the LeafNode in the path field MUST meet the same criteria as the LeafNode in an Update for the removed leaf (see Section 13.1.2). In particular, the credential in the LeafNode MUST present a set of identifiers that is acceptable to the application for the removed participant.

The PR moves it here (so it's only in one place).

This PR is editorial, so I'd rather not remove the restriction here but instead discuss this as a separate issue. I've also been wondering why only one Remove is allowed? E.g. I may want to remove a couple of my old devices.

Reviewer: Agree that we should handle the revalidation separately, in #660.

On "why only one" -- The idea is to *only* cover the resync case, where a device is replacing *its own* state.  Other devices can be cleaned up after the new member has joined.

  another LeafNode, the authentication service MUST additionally validate that
  the set of identities attested by the credential in the new LeafNode is
  acceptable relative to the identities attested by the old credential.

>   E.g. sending an Update proposal updates the sender's old LeafNode to a new
>   one, or joining via external commit updates the joiner's old LeafNode
>   included in the committed Remove proposal to a new one.

Reviewer: ```suggestion
  For example:
    * An Update proposal updates the sender's old LeafNode to a new one
    * A "resync" external commit removes the joiner's old LeafNode via a Remove proposal and replaces it with a new one
```

  specified by the `capabilities` field of each member's LeafNode, and that the
  `capabilities` field of this LeafNode indicates support for all the credential
  types currently in use by other members.

* Verify the `lifetime` field:
  * When validating a LeafNode in a KeyPackage before sending an Add proposal,
    the current time MUST be within the `lifetime` range.  A KeyPackage
    containing a LeafNode that is expired or not yet valid MUST NOT be sent in
    an Add proposal.
  * When receiving an Add or validating a tree, checking the `lifetime` is
    RECOMMENDED, if it is feasible in a given application context.  Because of
    the asynchronous nature of MLS, the `lifetime` may have been valid when the
    leaf node was proposed for addition, even if it is expired at these later
    points in the protocol.

* Verify that the `leaf_node_source` field has the appropriate value for the
  context in which the LeafNode is being validated (as defined in
  {{leaf-node-contents}}).

* Verify that the following fields in the LeafNode are unique among the
  members of the group (including any other members added in the same
  Commit):

    * `encryption_key`

>     * `signature_key`

Reviewer: Why are you removing this uniqueness requirement?

Reviewer: I see you have added it below.  That seems like a lot of repetition and risk of decoherence.  I would prefer to have it up here.

If the concern is the "added in the same Commit" part -- it seems like this requirement could be applied iteratively, in the sense that if you're updating/adding a bunch of leaves in a Commit and you do this check relative to the current tree before you apply an update/add, then that will catch a duplication.

Author: On second thought, I think this check belongs here.

Re. "added in the same Commit", I think a fix for this would be to change the "Validating Proposals" procedure to check individual validity at the end, after applying all committed proposals. I.e. the list is invalid if "After applying the proposal list, an individual proposal is invalid as specified in {{proposals}}."

Re. applying the check iteratively, the way I read the current draft, it seems that all checks are applied to the current state. For example, its valid to commit to A removing B and B removing A. But if we applied checks iteratively, it would be invalid.

>     of the `lifetime` field.

Reviewer: This description seems less clear to me than before.  Are you trying to be more generic?

Perhaps we should just have a generic RECOMMENDED here, and have the MUST in the Add section.

  extensions, proposals, and credential types MUST be listed in the LeafNode's
  `capabilities` field.

* Verify that the credential type is supported by all members of the group, as
  specified by the `capabilities` field of each member's LeafNode, and that the
  `capabilities` field of this LeafNode indicates support for all the credential
  types currently in use by other members.

* Verify the `lifetime` field:
  * If the LeafNode appears in a message being sent by the client, e.g., a
    proposal or a commit, then the client MUST verify that the current time is within
    the range of the `lifetime` field.
  * If instead the LeafNode appears in a message being received by the client, e.g.,
    a proposal, a commit, or a ratchet tree of the group the client is joining, it is
    RECOMMENDED that the client verifies that the current time is within the range
    of the `lifetime` field.

* Verify that the extensions in the LeafNode are supported by checking that the
  ID for each extension in the `extensions` field is listed in the
  `capabilities.extensions` field of the LeafNode.

* Verify the `leaf_node_source` field:

>   * If the LeafNode appears in a KeyPackage in an Add proposal, or in a downloaded
>     KeyPackage used to create an Add proposal, verify that `leaf_node_source` is
>     set to `key_package`.

Reviewer: ```suggestion
  * If the LeafNode appears in a KeyPackage, verify that `leaf_node_source` is
    set to `key_package`.
```

~~~

On receiving an MLSMessageContent containing a Proposal, a client MUST verify the
signature inside MLSMessageAuth and that the `epoch` field of the enclosing
MLSMessageContent is equal to the `epoch` field of the current GroupContext object.
If the verification is successful, then the Proposal should be cached in such a way
that it can be retrieved by hash (as a ProposalOrRef object) in a later Commit message.

### Add

An Add proposal requests that a client with a specified KeyPackage be added
to the group.

~~~ tls
struct {
    KeyPackage key_package;
} Add;
~~~

An Add proposal is invalid if any of the following is true:

* The KeyPackage is invalid according to {{keypackage-validation}}.

* The LeafNode in the KeyPackage shares the same `signature_key` with another
  LeafNode in the group.

* The LeafNode in the KeyPackage shares the same `encryption_key` with another
  LeafNode in the group.

* The Credential in the KeyPackage represents a client already in the

>   group according to the application.

Reviewer 2: I don't think this was agreed on

Author: This is how we interpreted the text
> If there are multiple Add proposals containing KeyPackages that the committer considers to represent the same client or a client already in the group (for example, identical KeyPackages or KeyPackages sharing the same Credential), the committer again chooses one to include and considers the rest invalid.

If this was not what was meant, the part "or a client already in the group" would be redundant. But of course I'll remove it if we were wrong.

~~~

On receiving an MLSMessageContent containing a Proposal, a client MUST verify the

> signature inside MLSMessageAuth and that the `epoch` field of the enclosing

Reviewer: We should delete these requirements here, and make them generic. It seems like you should always check the signature on MLSMessageContentAuth before using the content, regardless of the content type.

Author: Thanks for pointing this out! (We used to have a earlier version with more cases and more steps.)

} Update;
~~~

An Update proposal is invalid if any of the following is true:

> * The LeafNode is invalid for an Update proposal according to {{leaf-node-validation}}.

Reviewer: ```suggestion
```

that any members whose removal was proposed during the epoch are actually
removed before any application data is transmitted.

A sender and a receiver of a Commit MUST verify that the committed list of
proposals is valid as specified in {{validating-proposals}}. A list is invalid if, for example,

> it includes a proposal with an invalid signature, a proposal sent within a different epoch.

Reviewer: Suggest using a more interesting example here, such as including an Update and Remove for the same member.

or an Add when the sender does not have the application-level permission to add new users.

> The sender of a Commit SHOULD include all valid proposals that it has received
> during the current epoch, as long as this does not make the proposal list invalid.

Reviewer: ```suggestion
The sender of a Commit SHOULD include all proposals that it has received
during the current epoch, that are valid according to the rules for their
proposal types and according to application policy, as long as this results in
a valid proposal list.
```

Due to the asynchronous nature of proposals, receivers of a Commit SHOULD NOT enforce

> that all valid proposals sent within the current epoch are referenced by the next

Reviewer: FWIW I think "valid proposals" is a fine shorthand even with the above edit.

A member of the group creates a Commit message and the corresponding Welcome
message at the same time, by taking the following steps:

> * Verify that the committed list of proposals is valid as specified in
>   {{proposal-list-validation}}.

Reviewer: ```suggestion
* Verify that the list of proposals to be committed is valid as specified in
  {{proposal-list-validation}}.
```

root hash of the ratchet tree matches the `tree_hash` of the GroupContext before
using the tree for MLS operations.

> ## Proposal-List Validation

Reviewer: ```suggestion
## Proposal List Validation
```

Reviewer: I would put this above, probably as a subsection of {{proposals}}.  But we can handle that in a section reordering PR; I have added this to #672

root hash of the ratchet tree matches the `tree_hash` of the GroupContext before
using the tree for MLS operations.

## Validating Proposals

A group member creating a commit and a group member processing a commit
MUST verify that the list of committed proposals is valid using one of the following
procedures, depending on whether the commit is external or not.

For a regular, i.e. not external, commit the list is invalid if it contains any of the
following:

* An individual proposal that is invalid as specified in {{proposals}}.

* An Update proposal generated by the committer.

> * A Rmove proposal that removes the committer.

Reviewer 4: ```suggestion
* A Remove proposal that removes the committer.
```

  has received multiple such proposals they SHOULD prefer any Remove received, or
  the most recent Update if there are no Removes.

* Multiple Add proposals that contain KeyPackages that represent the same client
  according to the application (for example, identical KeyPackages or KeyPackages
  sharing the same Credential).

* Multiple PreSharedKey proposals that reference the same PreSharedKeyID.

* Multiple GroupContextExtensions proposals.

* A ReInit proposal together with any other proposal. If the committer has
  received other proposals during the epoch, they SHOULD prefer them over the
  ReInit proposal, allowing the ReInit to be resent and applied in a subsequent
  epoch.

* An ExternalInit proposal.

* A proposal with a non-default proposal type that is not supported by some
  members of the group that will process the Commit (i.e., members being added
  or removed by the Commit do not need to support the proposal type).

An application may extend the above procedure by additional rules, for example,
requiring application-level permissions to add members, or rules concerning
non-default proposal types.

> For an external commit, the list is valid if:

Reviewer: It's slightly confusing that the above is a negative check, and this is a positive check.  But that might be unavoidable.  I might rephrase as something like:

... the list is valid if it contains only the following proposals (not necessarily in this order):

* Exactly one ExternalInit proposal
* At most one Remove proposal...
* Zero or more PSK proposals...
* No other proposals

(The last bullet is a little redundant, but it bears repeating.)

Reviewer: (And maybe we should require order here!)
